base/shared/src/main/scala/scalaz/data/maybe2.scala

```diff
@@ -51,40 +52,41 @@ object Maybe2Module {
 }
 
 private[data] object Maybe2Impl extends Maybe2Module {
-  type Maybe2[A, B] = Option2[A, B]
+  type Maybe2[A, B] = Lazy[Option2[A, B]]
 
-  def just2[A, B](a: A, b: B): Maybe2[A, B] = Maybe2Impl.fromOption2(Some2(a, b))
-  def empty2[A, B]: Maybe2[A, B]            = None2
+  def just2[A, B](a: A, b: B): Maybe2[A, B]            = fromOption2(Some2(a, b))
+  def empty2[A, B]: Maybe2[A, B]                       = Lazy.now(None2)
+  def delay[A, B](m: () => Maybe2[A, B]): Maybe2[A, B] = Lazy.delay(m)
 
-  def isCovariant_1[B]: IsCovariant[Maybe2[?, B]] = IsCovariant.scalaCovariant[Option2[+?, B]]
-  def isCovariant_2[A]: IsCovariant[Maybe2[A, ?]] = IsCovariant.scalaCovariant[Option2[A, +?]]
+  def isCovariant_1[B]: IsCovariant[Maybe2[?, B]] = IsCovariant.unsafeForce // TODO: do better
+  def isCovariant_2[A]: IsCovariant[Maybe2[A, ?]] = IsCovariant.unsafeForce // TODO: do better
 
   def debug[A, B](implicit A: Debug[A], B: Debug[B]): Debug[Maybe2[A, B]] = {
     import Scalaz.debugInterpolator
     DebugClass.instance[Maybe2[A, B]] {
-      case Some2(a, b) => z"Just2($a, $b)"
-      case _           => Cord("Empty2")
+      case Lazy.Eval(Some2(a, b)) => z"Just2($a, $b)"
+      case _                      => Cord("Empty2")
     }
   }
 
   def eq[A, B](implicit A: Eq[A], B: Eq[B]): Eq[Maybe2[A, B]] =
     instanceOf[EqClass[Maybe2[A, B]]] {
-      case (Some2(a1, b1), Some2(a2, b2)) => A.equal(a1, a2) && B.equal(b1, b2)
-      case (None2, None2)                 => true
-      case _                              => false
+      case (Lazy.Eval(Some2(a1, b1)), Lazy.Eval(Some2(a2, b2))) => A.equal(a1, a2) && B.equal(b1, b2)
+      case (Lazy.Eval(None2), Lazy.Eval(None2))                 => true
+      case _                                                    => false
     }
 
   def bifunctor: Bifunctor[Maybe2] =
     instanceOf(new BifunctorClass[Maybe2] {
       override def bimap[A, B, S, T](fab: Maybe2[A, B])(as: A => S, bt: B => T): Maybe2[S, T] =
         fab match {
-          case Some2(_1, _2) => Some2(as(_1), bt(_2))
-          case None2         => None2
+          case Lazy.Eval(Some2(_1, _2)) => Lazy.now(Some2(as(_1), bt(_2)))
+          case _                        => Lazy.now(None2)
         }
     })
 
-  private[data] def fromOption2[A, B](o: Option2[A, B]): Maybe2[A, B] = o
-  private[data] def toOption2[A, B](m: Maybe2[A, B]): Option2[A, B]   = m
+  private[data] def fromOption2[A, B](o: Option2[A, B]): Maybe2[A, B] = Lazy.now(o)
+  private[data] def toOption2[A, B](m: Maybe2[A, B]): Option2[A, B]   = Lazy.eval(m)
```

Comment:
Yeah, no longer an isomorphism exactly. Guess this also means that whenever we use the hidden-variance trick we're making a strict backing type, so that's an inconsistency we're gonna have to live with if we use that trick + go with lazy-by-default data structures.

Reply:
If we ignore bottom values there are still isomorphic, though.